pwndbg/heap/ptmalloc.py

```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                        else:
+                            reg = ldrw_instr.operands[0].str
+                            if instr.mnemonic == 'add' and instr.op_str == reg + ", pc":
+                                # ldr.w reg, [pc, #offset]
+                                offset = ldrw_instr.operands[1].mem.disp
+                                offset = pwndbg.memory.s32((ldrw_instr.address + 4 & -4) + offset)
+                                # add reg, pc
+                                self._main_arena_addr = offset + instr.address + 4
+                else:
+                    inform_report_issue("main_arena")
+                    raise OSError("Cannot find the symbol via heuristics")
+            # try to search main_arena in .data of libc if we can't find it via above trick
+            if not self._main_arena_addr:
+                _IO_2_1_stdin_addr = pwndbg.symbol.address('_IO_2_1_stdin_')
+                _IO_list_all_addr = pwndbg.symbol.address('_IO_list_all')
+                # main_arena is between _IO_2_1_stdin and _IO_list_all
+                for addr in range(_IO_2_1_stdin_addr, _IO_list_all_addr, pwndbg.arch.ptrsize):
+                    tmp_arena = self.malloc_state(addr)
+                    if tmp_arena["next"] == addr:
+                        self._main_arena_addr = addr
+                        break
+                if not self._main_arena_addr:
+                    # there are more than one arena, try to find by main_arena.top and main_arena.max_system_mem
+                    heap_page = next(x for x in pwndbg.vmmap.get() if "heap]" in x.objfile)
+                    for addr in range(_IO_2_1_stdin_addr, _IO_list_all_addr, pwndbg.arch.ptrsize):
+                        tmp_arena = self.malloc_state(addr)
+                        if heap_page.start <= tmp_arena["top"] <= heap_page.end:
+                            if tmp_arena["max_system_mem"] != 0:
+                                self._main_arena_addr = addr
+                                break
+
+        if self._main_arena_addr:
+            self._main_arena = self.malloc_state(self._main_arena_addr)
+        else:
+            inform_report_issue("main_arena")
+            raise OSError("Cannot find the symbol via heuristics")
+
+        return self._main_arena
+
+    def has_tcache(self):
+        # TODO/FIXME: Can we determine the tcache_bins existence more reliable?
+
+        # There is no debug symbols, we determine the tcache_bins existence by checking glibc version only
+        return self.is_initialized() and pwndbg.glibc.get_version() >= (2, 26)
+
+    def get_tls_base_via_errno_location(self):
+        assert pwndbg.arch.current in ("x86-64", "i386", "arm")
+        # TODO/FIXME: Need a better way to find tls
+        already_lock = gdb.parameter("scheduler-locking") == "on"
+        if not already_lock:
+            gdb.execute("set scheduler-locking on")
+        errno_addr = int(gdb.parse_and_eval("(int *)__errno_location()"))
```

Comment:
Is there a reason we make the call and not fetch it by the `errno` symbol address?

Reply:
Maybe we don't have errno symbol address sometimes lol